plutus-core/plutus-core/src/PlutusCore/Builtin/Convert.hs

```diff
@@ -0,0 +1,539 @@
+-}
+
+{- Note [fromIntegral and padding arguments]
+
+As described in Note [How to add a built-in function: simple cases], we should
+not normally use a function like fromIntegral to convert arguments into more
+amenable forms, as it runs the risk of truncation. In the case of the padding
+argument given to integerToByteString (which is an Int) as opposed to how the
+builtin it implements is called (which passes an Integer), we use fromIntegral
+in spite of this. This is for two reasons: we are more-or-less forced to do so
+anyway (for reasons of efficiency if nothing else), and even if truncation
+does occur, we have much bigger issues. We elaborate on these points in this note.
+
+It is a fact that ByteString lengths are measured as Ints: while there are good
+reasons to disagree with this policy, it is forced on us, and can't really be
+changed. Furthermore, certain auxilliary tools we use to define
+integerToByteString (specifically, Builder) also uses Int to track its
+allocation size. As part of integerToByteString, we are in some cases required
+to compare one, or both, of these against the padding argument, possibly as
+often as once _per digit_. If we wanted to avoid fromIntegral, we would have
+to promote either the length of a ByteString, or the size of a Builder, to
+Integer, each time, only to then throw that value away. This would create
+unacceptable inefficiencies, not only due to the needless allocations (which
+would actually exceed the size of what is being generated by the function, as
+each Integer costs a minimum of _four_ machine words!), but also because
+Integer equality is far costlier (and far harder to optimize out) than Int
+equality. As an alternative, we could 'demote' the padding argument to an Int,
+but the only way we can do this is by using fromIntegral: this means that we
+have to do it _somewhere_, and the only real difference is location. By
+placing this conversion in a wrapper, we reduce the work the implementation
+has to do, and avoid cluttering it with yet more code.
+
+Furthermore, if truncation _would_ occur, then we have much bigger problems.
+On all supported platforms (which all use 64-bit Int representations),
+truncation in the padding argument would require passing an argument that is
+either larger than 2^62 - 1, or smaller than 2^63 - 1. In the first case,
+the cost of allocating something even half of that size would blow through any
+and all execution and memory limits long before anyone could even realize that
+the result they received is shorter than what they requested; in the second
+case, it wouldn't matter anyway, as no number that could reasonably fit on-chain
+would even come close to requiring such a number of digits at a minimum. In any
+reasonable situation, the padding argument would be much, _much_ smaller than
+the positive size limits for Int, and in cases where it isn't, the rest of the
+system prevents anything untoward being noticed.
```

Comment:
I'm confused about one case: if we remove the current arbitrary limit, and I pass an `Integer` which wraps around to a positive `Int` that is reasonably small... then currently it looks like that will succeed? Are we okay with that? We would then be relying on costing to rule out any excessively large padding arguments.

Reply:
This has now been fixed, since we no longer accept negative length requests.